features/202-session-management/design.md

```diff
@@ -0,0 +1,163 @@
+# Background
+
+Sessions of auth gear should be configurable and manageable.
+
+
+# Objectives
+
+- App Developer/User should be able to manage sessions.
+- App Developer should be able to configure session lifetime.
+- App Developer should be able to configure session token transport.
+
+
+# Proposed Design
+
+## Stateful Session
+
+Sessions would be stateful (i.e. persisted into database), instead of being
+stateless.
+
+Each session would be associated with attributes:
+- session ID
+- app ID
+- user ID
+- identity ID
+- expiration time
+- creation time
+- last access time
+- creation IP
+- last IP
+- creation user agent
+- last user agent
+- custom attributes
+
+Session attributes would be available in web-hook event context, and session
+management APIs.
```

Comment:
Will microservices be able to access the context from header too?

Reply:
oh I missed this sentence when I first read the draft.

Do we need to have the session attributes in context actually? Is there some usecases that made @kiootic think they should be?

Reply:
I don't think so, it would bloat the request quite a lot. Quick testing shows that including session attributes as plain JSON object would make request size +70% (+20% gzipped). I think I'll include a get & update session attribute API instead and set just the session ID in header.

Reply:
So we shouldn't have it in web-hook event context too?

Reply:
Web-hook events are generated far less frequently than every request. Also, developer need to opt-in to which events to generate.

Reply:
But is there a frequent use case webhook event needs session data to justify the cost involved?

Reply:
After further consideration, web-hook handler would rather want just some of session attributes, e.g. check user agent/IP on sign up. I'll remove the session attributes (except session ID) from web-hook event context, and developer can get the attributes using API if required.

Reply:
noted, actually if you think there are good reasons I'm okay if we just inject the whole session in it... just want to make sure it is considered carefully (coz it was not explained)